drivers/spi/spi_pico_pio.c

```diff
@@ -0,0 +1,415 @@
+			    0x5101, /*  1: in     pins, 1         side 1 [1] */
+				/*     .wrap */
+);
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_1_cpha_1, 0, 2,
+			    /*     .wrap_target */
+			    0x7021, /*  0: out    x, 1            side 1 */
+			    0xa101, /*  1: mov    pins, x         side 0 [1] */
+			    0x5001, /*  2: in     pins, 1         side 1 */
+				/*     .wrap */
+);
+
+static float spi_pico_pio_clock_divisor(const struct spi_pico_pio_config *dev_cfg,
+	uint32_t spi_frequency)
+{
+	return (float)dev_cfg->clock_freq / (float)(PIO_CYCLES * spi_frequency);
+}
+
+static uint32_t spi_pico_pio_maximum_clock_frequency(const struct spi_pico_pio_config *dev_cfg)
+{
+	return dev_cfg->clock_freq / PIO_CYCLES;
+}
+
+static uint32_t spi_pico_pio_minimum_clock_frequency(const struct spi_pico_pio_config *dev_cfg)
+{
+	return dev_cfg->clock_freq / (PIO_CYCLES * 65536);
+}
+
+static inline bool spi_pico_pio_transfer_ongoing(struct spi_pico_pio_data *data)
+{
+	return spi_context_tx_on(&data->spi_ctx) || spi_context_rx_on(&data->spi_ctx);
+}
+
+/* TODO:  Add support for 16- and 32-bit writes */
+static inline void spi_pico_pio_sm_put8(PIO pio, uint sm, uint8_t data)
+{
+	/* Do 8 bit accesses on FIFO, so that write data is byte-replicated. This */
+	/* gets us the left-justification for free (for MSB-first shift-out) */
+	io_rw_8 *txfifo = (io_rw_8 *)&pio->txf[sm];
+
+	*txfifo = data;
+}
+
+/* TODO:  Add support for 16- and 32-bit reads */
+static inline uint8_t spi_pico_pio_sm_get8(PIO pio, uint sm)
+{
+	/* Do 8 bit accesses on FIFO, so that write data is byte-replicated. This */
+	/* gets us the left-justification for free (for MSB-first shift-out) */
+	io_rw_8 *rxfifo = (io_rw_8 *)&pio->rxf[sm];
+
+	return *rxfifo;
+}
+
+static int spi_pico_pio_configure(const struct spi_pico_pio_config *dev_cfg,
+				  struct spi_pico_pio_data *data, const struct spi_config *spi_cfg)
+{
+	const struct gpio_dt_spec *miso = NULL;
+	const struct gpio_dt_spec *mosi = NULL;
+	const struct gpio_dt_spec *clk = NULL;
+	pio_sm_config sm_config;
+	uint32_t offset;
+	uint32_t wrap_target;
+	uint32_t wrap;
+	const pio_program_t *program;
+	int rc;
+
+	if (spi_context_configured(&data->spi_ctx, spi_cfg)) {
+		return 0;
+	}
+
+	if (spi_cfg->operation & SPI_OP_MODE_SLAVE) {
+		LOG_ERR("Slave mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (spi_cfg->operation & SPI_TRANSFER_LSB) {
+		LOG_ERR("Unsupported configuration");
+		return -ENOTSUP;
+	}
+
+#if defined(CONFIG_SPI_EXTENDED_MODES)
+	if (spi_cfg->operation & (SPI_LINES_DUAL | SPI_LINES_QUAD | SPI_LINES_OCTAL)) {
+		LOG_ERR("Unsupported configuration");
+		return -ENOTSUP;
+	}
+#endif /* CONFIG_SPI_EXTENDED_MODES */
+
+	const int bits = SPI_WORD_SIZE_GET(spi_cfg->operation);
+
+	if (bits != 8) {
+		LOG_ERR("Only 8 bit word size is supported");
+		return -ENOTSUP;
+	}
+
+	data->bits = bits;
+	data->dfs = ((data->bits - 1) / 8) + 1;
+
+	if ((spi_cfg->frequency > spi_pico_pio_maximum_clock_frequency(dev_cfg)) ||
```

Comment:
It should better compare with `>=`.
Intuitively, it seems better to reverse the left hand side and right hand side of `||`.

Reply:
The maximum and minimum frequency values are allowed, hence the use of "<" and ">".  But you're right about the order of the comparisons.